lymph/midline.py

```diff
@@ -758,17 +502,70 @@ def likelihood(
         except ValueError:
             return -np.inf if log else 0.
 
-        if len(self.patient_data.loc[(self.patient_data['info', 'tumor', 'midline_extension']==False) | (self.patient_data['info', 'tumor', 'midline_extension']==True)])>0:
+        llh = 0. if log else 1.
+
+        stored_t_stages = set(self.ext.ipsi.diagnose_matrices.keys())
+        provided_t_stages = set(self.ext.ipsi.diag_time_dists.keys())
+        t_stages = list(stored_t_stages.intersection(provided_t_stages))
+        
+        max_t = self.diag_time_dists.max_t
+        llh = 0. if log else 1.
+
+        state_probs_ipsi_nox = self.noext.ipsi._evolve_onestep()
+        state_probs_contra_nox = self.noext.contra._evolve_onestep()
+        state_probs_midext = self._evolve_midext(new_params = given_params)
+        state_probs_ipsi_ex = np.zeros(
+        shape=(max_t + 1, len(self.ext.ipsi.state_list)),
+        dtype=float
+        )
+        state_probs_ipsi_ex[0,0] = 1.
+        state_probs_contra_ex = np.zeros(
+        shape=(max_t + 1, len(self.ext.ipsi.state_list)),
+        dtype=float
+        )
+        state_probs_contra_ex[0,0] = 1.
+
+        for stage in t_stages:
+            for i in range(max_t):
+                state_probs_ipsi_ex[(i+1),:] = state_probs_midext[i,1] * state_probs_ipsi_ex[i,:] @ self.ext.ipsi.transition_matrix + state_probs_midext[i,0] * state_probs_ipsi_nox[(i+1),:]
+                state_probs_contra_ex[(i+1),:] = state_probs_midext[i,1] * state_probs_contra_ex[i,:] @ self.ext.contra.transition_matrix + state_probs_midext[i,0] * state_probs_contra_nox[(i+1),:]
+
+            joint_state_probs_nox = (
+                state_probs_ipsi_nox.T
+                @ np.diag(self.noext.ipsi.diag_time_dists[stage].pmf)
+                @ state_probs_contra_nox
+            )
+            joint_state_probs_ex = (
+                state_probs_ipsi_ex.T
+                @ np.diag(self.ext.ipsi.diag_time_dists[stage].pmf)
+                @ state_probs_contra_ex
+            )
+            p_ex = np.sum(
+                self.ext.ipsi.diagnose_matrices[stage]
+                * (joint_state_probs_ex
+                    @ self.ext.contra.diagnose_matrices[stage]),
+                axis=0
+                )
+            p_nox = np.sum(
+                self.noext.ipsi.diagnose_matrices[stage]
+                * (joint_state_probs_nox
+                    @ self.noext.contra.diagnose_matrices[stage]),
+                axis=0
+                )
+
+            p = np.array([p_nox,p_ex]) * np.array([[state_probs_midext[max_t,0]], [state_probs_midext[max_t,1]]])
+
             if log:
-                llh += self.ext._likelihood_mid(log=log)
-                llh += self.noext._likelihood_mid(log=log)
+                p = np.log(p)
+                llh_nox_ex = p @ self.diagnose_matrices_midext[stage]
+                llh += np.sum(llh_nox_ex)
+            #probably wrong. Matrix multiplication sums over patients but it should multiply.
```

Comment:
I think it's the other way around: In the `log` case, you should first matrix-multiply with the midline extension diagnose matrix and after that take the logarithm... But I might be wrong.